wgengine/tsdns/tsdns.go

```diff
@@ -0,0 +1,274 @@
+	r := &Resolver{
+		logf: logf,
+		ip:   defaultIP,
+		port: defaultPort,
+	}
+
+	return r
+}
+
+// AcceptsPacket determines if the given packet is
+// directed to this resolver (by ip and port).
+// We also require that UDP be used to simplify things for now.
+func (r *Resolver) AcceptsPacket(in *packet.ParsedPacket) bool {
+	return in.DstIP == r.ip && in.DstPort == r.port && in.IPProto == packet.UDP
+}
+
+// SetMap sets the resolver's DNS map.
+func (r *Resolver) SetMap(m *Map) {
+	r.mu.Lock()
+	r.dnsMap = m
+	r.mu.Unlock()
+}
+
+// Resolve maps a given domain name to the IP address of the host that owns it.
+func (r *Resolver) Resolve(domain string) (netaddr.IP, dns.RCode, error) {
+	// If not a subdomain of ipn.dev, then we must refuse this query.
+	// We do this before checking the map to distinguish beween nonexistent domains
+	// and misdirected queries.
+	if !strings.HasSuffix(domain, ".ipn.dev") {
+		return netaddr.IP{}, dns.RCodeRefused, errNotOurName
+	}
+
+	r.mu.Lock()
+	if r.dnsMap == nil {
+		r.mu.Unlock()
+		return netaddr.IP{}, dns.RCodeServerFailure, errMapNotSet
+	}
+	addr, found := r.dnsMap.DomainToIP[domain]
+	r.mu.Unlock()
+
+	if !found {
+		return netaddr.IP{}, dns.RCodeNameError, errNoSuchDomain
+	}
+	return addr, dns.RCodeSuccess, nil
+}
+
+type response struct {
+	Header         dns.Header
+	ResourceHeader dns.ResourceHeader
+	Question       dns.Question
+	// TODO(dmytro): support IPv6.
+	IP netaddr.IP
+}
+
+// parseQuery parses the query in given packet into a response struct.
+func (r *Resolver) parseQuery(query *packet.ParsedPacket, resp *response) error {
+	var parser dns.Parser
+	var err error
+
+	resp.Header, err = parser.Start(query.Payload())
+	if err != nil {
+		return err
+	}
+
+	if resp.Header.Response {
+		return errNotQuery
+	}
+
+	resp.Question, err = parser.Question()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// makeResponse resolves the question stored in resp and sets the answer fields.
+func (r *Resolver) makeResponse(resp *response) error {
+	var err error
+
+	name := resp.Question.Name.String()
+	if len(name) > 0 {
+		name = name[:len(name)-1]
+	}
+
+	if resp.Question.Type == dns.TypeA {
+		// Remove final dot from name: *.ipn.dev. -> *.ipn.dev
+		resp.IP, resp.Header.RCode, err = r.Resolve(name)
+	} else {
+		resp.Header.RCode = dns.RCodeNotImplemented
+		err = errNotImplemented
+	}
+
+	return err
+}
+
+// marshalAnswer serializes the answer record into an active builder.
+func marshalAnswer(resp *response, builder *dns.Builder) error {
+	var answer dns.AResource
+
+	err := builder.StartAnswers()
+	if err != nil {
+		return err
+	}
+
+	answerHeader := dns.ResourceHeader{
+		Name:  resp.Question.Name,
+		Type:  dns.TypeA,
+		Class: dns.ClassINET,
+		TTL:   defaultTTL,
+	}
+	ip := resp.IP.As16()
+	copy(answer.A[:], ip[12:])
+	return builder.AResource(answerHeader, answer)
+}
+
+// marshalResponse serializes the DNS response into an active builder.
```

Comment:
you don't say what it returns. But why does this func call finish if caller passed it in? might the caller keep building with it?